File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.swift
// 문제 분석
// 완전탐색으로 풀이하면 2 ** 20 ~= 1,000,000 정도여서 사실 가능하다. 그러나 그래프 탐색으로 풀어보자.
// 모든 경우의 수를 따지려면 DFS가 적합하겠다
// 방향이 두 가지(+, -)가 있는 그래프를 탐색한다고 생각하면 쉽다
// dfs 함수가 재귀적으로 + / - 의 경우를 각각 호출하고, 
// 깊이가 numbers와 같아지면, target과 비교하여 cnt하는 방식으로 풀어보자

// 의사 코드
// func dfs(depth:int, sum:Int) -> Int
    // depth == numbers.size && sum == target인 경우 cnt += 1
    // depth + 1이 범위(numbers.size) 벗어나면 return
    // 두 가지 방향 재귀적으로 탐색
    // dfs(depth + 1, sum + numbers[depth + 1])
    // dfs(depth + 1, sum - numbers[depth + 1])
// func solution()
// dfs(-1, 0) 호출
// return cnt

import Foundation

func solution(_ numbers:[Int], _ target:Int) -> Int {
    var cnt = 0
    
    func dfs(_ depth: Int, _ sum: Int) {
        // depth == numbers.count - 1 && sum == target인 경우 cnt += 1
        if depth == numbers.count - 1 && sum == target {
            cnt += 1
            return // 재귀 탐색 종료
        }
        // depth + 1이 범위(numbers.count)보다 작으면 진행, 벗어나면 return
        guard depth + 1 < numbers.count else { return }
        // 두 가지 방향 재귀적으로 탐색
        // dfs(depth + 1, sum + numbers[depth + 1])
        dfs(depth + 1, sum + numbers[depth + 1])
        // dfs(depth + 1, sum - numbers[depth + 1])
        dfs(depth + 1, sum - numbers[depth + 1])
    }
    
    // dfs(-1, 0) 호출
    dfs(-1, 0) // 첫 번째 인덱스부터 방향이 갈리므로
    
    // return cnt
    return cnt
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.py
# 이해
# numbers에 주어진 원소를 더하거나 빼서 target을 완성시키는 방법의 수

# 풀이
# DFS로 모든 가능성을 탐색한다.
# 2 ** 20 ~= 1,000,000 이므로 문제 없다.
# 2 ** 30 ~= 1,000,000,000이므로 2 ** 29까지는 문제 없을 것이다. O(N) 기준.

# numbers를 탐색한다.
# 1. number를 더하기 한 것을 재귀적으로 호출한다.
# 2. number를 빼기 한 것을 재귀적으로 호출한다.
# cnt == target이 되면 answer += 1
# answer를 return

cnt = 0

def solution(numbers, target):
    dfs(numbers, target, 0, 0)
    return cnt

def dfs(numbers, target, current, idx):
    global cnt
    if idx == len(numbers):
        if current == target:
            cnt += 1
        return
    
    dfs(numbers, target, current + numbers[idx], idx + 1)
    dfs(numbers, target, current - numbers[idx], idx + 1)
    
    
    
    
    
    
    
    
    
    
    